### PR TITLE
alphametics: Revised, increased speed

### PR DESCRIPTION
**Pull request by nikamirrr**
Reworked the solution, eliminated few inefficiencies

### CODE REVIEW

File: exercises/alphametics/example.py [hunk 1]
```diff
@@ -1,42 +1,99 @@
...
+        # Generate addtional non-zero digit permutations
+        for lonz in permutations(remnzdigs, len(lonzdigits)):
+            # Build a dictionary for to test the expression
+            t_digvals = lorest + lonz
+            # Evaluate the expression sides
+            testsum = sum([dig * loexpdict[let]
+                           for let, dig in zip(t_lowdigs, t_digvals)])
+            if testsum % 10 == 0:
+                # Low digit test passed, check the main digits
+                # print(lowdigdict)
```
Comment (cmccandless) [inline exercises/alphametics/example.py]: Please remove `print()` statements (even commented out).

**Comment by nikamirrr** [comment]
I think we can do even better, like a recursive algorithm to check the lowest digits,
move to next digits, generate them and check them.
A carryover will be accumulated for the next digits.
I am going to try to implement it over the weekend. Yesterday attempt was not very successful.
I may also try to design a special test case, which will cripple down the inefficient code.

**Comment by nikamirrr** [comment]
Since the pull request is approved. Is there anything else I need to do on my side? Please, let me know

**Comment by cmccandless** [comment]
@nikamirrr Regarding a recursive solution, it may not be as fast as you think. From [this article by Lars Kruse (found on the Python Wiki)](https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Data_Aggregation):
> Function call overhead in Python is relatively high, especially compared with the execution speed of a builtin function.

**Comment by nikamirrr** [comment]
I understand the general impact of the recursion. But if it allows to drop bad paths early enough it will be beneficial

**Comment by cmccandless** [comment]
@nikamirrr Fair enough. I'm curious to see how it turns out.

**Comment by N-Parsons** [comment]
Looks like run time is down to 0.5 seconds now compared to ~5 seconds with the previous iteration.

**Comment by nikamirrr** [comment]
Please, merge as is. Any further improvement is experimental and a bit less likely.

**Comment by cmccandless** [comment]
@nikamirrr Merged. Thanks!

**Comment by nikamirrr** [comment]
@cmccandless 
Looks like the recursive solution is faster.
The function call overhead is compensated by the fact that a good found candidate is traced through the next levels immediately and bad ones are rejected faster and do not build up.

**Pull request merged**